File: Depth-First Search/Binary tree/Find paths/LET_no337_House Robber III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: TreeNode) -> int:
        cur_not_used, cur_used = self.helper(root)
        return max(cur_not_used, cur_used)
    
    def helper(self, root):
        if not root:
            return 0, 0
        
        left_not_used, left_used = self.helper(root.left)
        right_not_used, right_used = self.helper(root.right)
        
        cur_not_used = max(left_used, left_not_used) + max(right_used, right_not_used)
        cur_used = left_not_used + right_not_used + root.val
        
        return cur_not_used, cur_used
        

File: Depth-First Search/Binary tree/Find paths/Medium/LET_no337_House Robber III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        int curLeft = 0, curRight = 0;
        return helper(root, curLeft, curRight);
    }
    
    int helper(TreeNode* cur, int& curLeft, int& curRight) {
        if (!cur)
            return 0;
            
        int leftLeft = 0, leftRight = 0, rightLeft = 0, rightRight = 0;
        curLeft = helper(cur->left, leftLeft, leftRight);
        curRight = helper(cur->right, rightLeft, rightRight);
        
        return max(cur->val + leftLeft + leftRight + rightLeft + rightRight, curLeft + curRight);
    }
};